rclc/src/functions.c

```diff
@@ -0,0 +1,255 @@
+    }
+    else if(rc != RCL_RET_OK) {
+      fprintf(stderr, "[rclc_spin_node] error in rcl_wait: %s\n", rcl_get_error_string_safe());
+      rcl_reset_error();
+    }
+
+    for (size_t i = 0; i < wait_set.size_of_subscriptions; ++i) {
+      if (wait_set.subscriptions[i]) {
+        char msg[64];
+        rc = rcl_take(wait_set.subscriptions[i], (void*)&msg, NULL);
+        if(rc != RCL_RET_OK) {
+          fprintf(stderr, "[rclc_spin_node] error in rcl_take: %s\n", rcl_get_error_string_safe());
+          rcl_reset_error();
+        }
+        else {
+          for (size_t i = 0; i < node->subs_s; ++i) {
+            if(&node->subs[i]->rcl_subscription == wait_set.subscriptions[i]) {
+              node->subs[i]->user_callback((const void*)&msg);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  rc = rcl_wait_set_fini(&wait_set);
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_spin_node] error in rcl_wait_set_fini: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+}
+
+rclc_node_t * rclc_create_node(const char * name)
+{
+  const char* namespace_ = ""; // TODO : put as parameter
+
+  rclc_node_t* ret = ALLOCATE(sizeof(rclc_node_t));
+  ret->rcl_node = rcl_get_zero_initialized_node();
+  ret->subs = NULL;
+  ret->subs_s = 0;
+  
+  rcl_node_options_t node_ops = rcl_node_get_default_options();
+  rcl_ret_t rc = rcl_node_init(&ret->rcl_node, name, namespace_, &node_ops);
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_create_node] error in rcl_node_init: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+    return NULL;
+  }
+  
+  return ret;
+}
+
+rclc_ret_t rclc_destroy_node(rclc_node_t * node)
+{
+  rcl_ret_t ret = rcl_node_fini(&node->rcl_node);
+  DEALLOCATE(node);
+  return ret;
+}
+
+rclc_publisher_t * rclc_create_publisher(
+  rclc_node_t * node,
+  const rosidl_message_type_support_t * type_support,
+  const char * topic_name,
+  size_t queue_size)
+{
+  (void)queue_size;
+  
+  rclc_publisher_t* ret = ALLOCATE(sizeof(rclc_publisher_t));
+  ret->rcl_publisher = rcl_get_zero_initialized_publisher();
+  ret->node = node;
+
+  rcl_publisher_options_t pub_opt = rcl_publisher_get_default_options();
+  rcl_ret_t rc = rcl_publisher_init(&ret->rcl_publisher, &node->rcl_node, type_support, topic_name, &pub_opt);
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_create_publisher] error in rcl_publisher_init: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+    return NULL;
+  }
+
+  return ret;
+}
+
+rclc_ret_t rclc_destroy_publisher(rclc_publisher_t * publisher)
+{
+  rcl_ret_t rc = rcl_publisher_fini(&publisher->rcl_publisher, &publisher->node->rcl_node);
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_destroy_publisher] error in rcl_publisher_fini: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+  DEALLOCATE(publisher);
+  return RCL_RET_OK;
+}
+
+rclc_ret_t rclc_publish(const rclc_publisher_t * publisher, const void * ros_message)
+{
```

Comment:
this should check if `publisher` is `NULL` before trying to access `&publisher->rcl_publisher`